# Выпускной проект по SQL: анализ сервиса для чтения книг по подписке ##

****Описание данных****

**Таблица books**

**book_id** — идентификатор книги

**author_id** — идентификатор автора

**title** — название книги

**num_pages** — количество страниц

**publication_date** — дата публикации книги

**publisher_id** — идентификатор издателя

**Таблица authors**

**author_id** — идентификатор автора

**author** — имя автора

**Таблица publishers**

**publisher_id** — идентификатор издательства

**publisher** — название издательства

**Таблица ratings**

**rating_id** — идентификатор оценки

**book_id** — идентификатор книги

**username** — имя пользователя, оставившего оценку

**rating** — оценка книги

**Таблица reviews**

**review_id** — идентификатор обзора

**book_id** — идентификатор книги

**username** — имя автора обзора

**text** — текст обзора

**Задания**
1. Посчитайте, сколько книг вышло после 1 января 2000 года;

2. Для каждой книги посчитайте количество обзоров и среднюю оценку;

3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [21]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [22]:
query = '''SELECT * FROM authors LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [23]:
query = '''SELECT * FROM publishers LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [24]:
query = '''SELECT * FROM ratings LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [25]:
query = '''SELECT * FROM reviews LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

**1. Количество книг, вышедших после 1 января 2000 года**

In [26]:
query = '''SELECT COUNT (book_id)
           FROM books
           WHERE publication_date > '2000-01-01'
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

После 1 января 2020 года выпущено 819 книг.

**2.Количество обзоров и средняя оценка для каждой книги**

In [27]:
query = '''SELECT b.title, COUNT (DISTINCT r.review_id) AS snt_review, AVG (rat.rating) AS avg_rating
           FROM books AS b
           INNER JOIN reviews AS r ON b.book_id = r.book_id
           INNER JOIN ratings AS rat ON b.book_id = rat.book_id
           GROUP BY b.title
           ORDER BY avg_rating DESC
           LIMIT 50
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

title  snt_review  avg_rating
0                   Misty of Chincoteague (Misty  #1)           2    5.000000
1           A Woman of Substance (Emma Harte Saga #1)           2    5.000000
2   The War of Art: Break Through the Blocks & Win...           2    5.000000
3                                       Evening Class           2    5.000000
4   Wherever You Go  There You Are: Mindfulness Me...           2    5.000000
5                                          Dead Souls           2    5.000000
6                            Tai-Pan (Asian Saga  #2)           2    5.000000
7   The Ghost Map: The Story of London's Most Terr...           2    5.000000
8                                         Angels Fall           2    5.000000
9                                       Alas  Babylon           2    5.000000
10                          Triptych (Will Trent  #1)           2    5.000000
11                                 The Woman in Black           2    5.000000
12               Welcome to Temptation (Dempseys  #1)           2    5.000000
13  Stone of Farewell (Memory  Sorrow  and Thorn  #2)           1    5.000000
14                                    Light in August           2    5.000000
15                           Neil Gaiman's Neverwhere           2    5.000000
16  The Demon-Haunted World: Science as a Candle i...           2    5.000000
17                 The Big Bad Wolf  (Alex Cross  #9)           2    5.000000
18                                        Marvel 1602           2    5.000000
19                                              Women           2    5.000000
20                               My Name Is Asher Lev           2    5.000000
21       Moneyball: The Art of Winning an Unfair Game           3    5.000000
22               Piercing the Darkness (Darkness  #2)           2    5.000000
23               Pop Goes the Weasel (Alex Cross  #5)           2    5.000000
24                       Plum Island (John Corey  #1)           2    5.000000
25                      Emily of New Moon (Emily  #1)           2    5.000000
26  The Adventures of Tom Sawyer and Adventures of...           1    5.000000
27            School's Out—Forever (Maximum Ride  #2)           3    5.000000
28           Evil Under the Sun (Hercule Poirot  #24)           2    5.000000
29               Stolen (Women of the Otherworld  #2)           2    5.000000
30                                Homage to Catalonia           2    5.000000
31                                          Geek Love           2    5.000000
32                      A Dirty Job (Grim Reaper  #1)           4    5.000000
33                                         Hard Times           2    5.000000
34  In the Hand of the Goddess (Song of the Liones...           2    5.000000
35                    Act of Treason (Mitch Rapp  #9)           2    5.000000
36      Arrows of the Queen (Heralds of Valdemar  #1)           2    5.000000
37  How to Be a Domestic Goddess: Baking and the A...           1    5.000000
38              A Fistful of Charms (The Hollows  #4)           2    5.000000
39  Crucial Conversations: Tools for Talking When ...           2    5.000000
40                                              March           2    5.000000
41  The Walking Dead  Book One (The Walking Dead #...           2    5.000000
42  Captivating: Unveiling the Mystery of a Woman'...           2    5.000000
43                 Dragonfly in Amber (Outlander  #2)           4    4.875000
44             Preludes & Nocturnes (The Sandman  #1)           3    4.833333
45  The Witching Hour (Lives of the Mayfair Witche...           3    4.750000
46      Lover Revealed (Black Dagger Brotherhood  #4)           3    4.750000
47                The Monster at the End of this Book           3    4.750000
48                                          The Twits           3    4.750000
49                                    House of Leaves           4    4.750000

У 43 книг рейтинг равен 5, но количество обзоров у большинства из них 2. Больше всего отзывов (4) с рейтингом 5.0 у книги A Dirty Job (Grim Reaper #1)

In [38]:
query = '''SELECT b.title, COUNT (DISTINCT r.review_id) AS snt_review, AVG (rat.rating) AS avg_rating
           FROM books AS b
           LEFT JOIN reviews AS r ON b.book_id = r.book_id
           LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
           GROUP BY b.book_id
           
           ORDER BY avg_rating DESC
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

title  snt_review  avg_rating
0        Arrows of the Queen (Heralds of Valdemar  #1)           2        5.00
1    The Walking Dead  Book One (The Walking Dead #...           2        5.00
2                                      Light in August           2        5.00
3    Wherever You Go  There You Are: Mindfulness Me...           2        5.00
4    Captivating: Unveiling the Mystery of a Woman'...           2        5.00
..                                                 ...         ...         ...
995  The World Is Flat: A Brief History of the Twen...           3        2.25
996                  His Excellency: George Washington           2        2.00
997                                      Drowning Ruth           3        2.00
998                                              Junky           2        2.00
999                               Harvesting the Heart           2        1.50

[1000 rows x 3 columns]

In [39]:
query = '''SELECT b.title, COUNT (DISTINCT r.review_id) AS snt_review, AVG (rat.rating) AS avg_rating
           FROM books AS b
           LEFT JOIN reviews AS r ON b.book_id = r.book_id
           LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
           GROUP BY b.book_id
           HAVING AVG (rat.rating) = 5.0
           ORDER BY snt_review DESC
           
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

title  snt_review  avg_rating
0                       A Dirty Job (Grim Reaper  #1)           4         5.0
1             School's Out—Forever (Maximum Ride  #2)           3         5.0
2        Moneyball: The Art of Winning an Unfair Game           3         5.0
3                     Act of Treason (Mitch Rapp  #9)           2         5.0
4                                       Alas  Babylon           2         5.0
5                                         Angels Fall           2         5.0
6       Arrows of the Queen (Heralds of Valdemar  #1)           2         5.0
7   Captivating: Unveiling the Mystery of a Woman'...           2         5.0
8   Crucial Conversations: Tools for Talking When ...           2         5.0
9                                          Dead Souls           2         5.0
10                      Emily of New Moon (Emily  #1)           2         5.0
11                                      Evening Class           2         5.0
12           Evil Under the Sun (Hercule Poirot  #24)           2         5.0
13                                          Geek Love           2         5.0
14                                         Hard Times           2         5.0
15                                Homage to Catalonia           2         5.0
16  In the Hand of the Goddess (Song of the Liones...           2         5.0
17                                    Light in August           2         5.0
18                                              March           2         5.0
19                                        Marvel 1602           2         5.0
20                  Misty of Chincoteague (Misty  #1)           2         5.0
21                               My Name Is Asher Lev           2         5.0
22                           Neil Gaiman's Neverwhere           2         5.0
23               Piercing the Darkness (Darkness  #2)           2         5.0
24                       Plum Island (John Corey  #1)           2         5.0
25               Pop Goes the Weasel (Alex Cross  #5)           2         5.0
26               Stolen (Women of the Otherworld  #2)           2         5.0
27                                              Women           2         5.0
28              A Fistful of Charms (The Hollows  #4)           2         5.0
29          A Woman of Substance (Emma Harte Saga #1)           2         5.0
30                                 The Woman in Black           2         5.0
31                          Triptych (Will Trent  #1)           2         5.0
32               Welcome to Temptation (Dempseys  #1)           2         5.0
33  Wherever You Go  There You Are: Mindfulness Me...           2         5.0
34                           Tai-Pan (Asian Saga  #2)           2         5.0
35                 The Big Bad Wolf  (Alex Cross  #9)           2         5.0
36  The Demon-Haunted World: Science as a Candle i...           2         5.0
37  The Ghost Map: The Story of London's Most Terr...           2         5.0
38  The Walking Dead  Book One (The Walking Dead #...           2         5.0
39  The War of Art: Break Through the Blocks & Win...           2         5.0
40  Stone of Farewell (Memory  Sorrow  and Thorn  #2)           1         5.0
41  The Adventures of Tom Sawyer and Adventures of...           1         5.0
42  How to Be a Domestic Goddess: Baking and the A...           1         5.0
43   The Cat in the Hat and Other Dr. Seuss Favorites           0         5.0

In [32]:
query = '''SELECT b.title, COUNT (DISTINCT r.review_id) AS snt_review, AVG (rat.rating) AS avg_rating
           FROM books AS b
           LEFT JOIN reviews AS r ON b.book_id = r.book_id
           LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
           GROUP BY b.book_id
           ORDER BY snt_review DESC
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

title  snt_review  avg_rating
0                              Twilight (Twilight  #1)           7    3.662500
1                                  Water for Elephants           6    3.977273
2                                     The Glass Castle           6    4.206897
3    Harry Potter and the Prisoner of Azkaban (Harr...           6    4.414634
4    The Curious Incident of the Dog in the Night-Time           6    4.081081
..                                                 ...         ...         ...
995    Anne Rice's The Vampire Lestat: A Graphic Novel           0    3.666667
996                            The Natural Way to Draw           0    3.000000
997   The Cat in the Hat and Other Dr. Seuss Favorites           0    5.000000
998                          Essential Tales and Poems           0    4.000000
999  Disney's Beauty and the Beast (A Little Golden...           0    4.000000

[1000 rows x 3 columns]

У 44 книг рейтинг равен 5, но количество обзоров у большинства из них 2. Больше всего отзывов (4) с рейтингом 5.0 у книги A Dirty Job (Grim Reaper #1). Больше всего отзывов у книги Twilight (Twilight #1)	 - 7.

**3.Идательство, выпустившее наибольшее число книг толще 50 страниц**

In [33]:
query = '''SELECT p.publisher, COUNT (DISTINCT book_id) AS snt_books
           FROM publishers AS p
           LEFT JOIN books AS b ON p.publisher_id = b.publisher_id
           WHERE b.num_pages > 50
           GROUP BY p.publisher
           ORDER BY snt_books DESC
           LIMIT 1
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  snt_books
0  Penguin Books         42

Penguin Books - издательство, выпустившее наибольшее количество книг (42) толщиной более 50 страниц.

**4.Автор с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками**

In [34]:
query = '''SELECT a.author, AVG (rat.rating) AS avg_rating, COUNT (DISTINCT rat.rating_id) AS snt_rating
           FROM authors AS a
           LEFT JOIN books AS b ON a.author_id = b.author_id
           LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
           GROUP BY a.author
           HAVING COUNT (DISTINCT rat.rating_id) >= 50
           
           ORDER BY avg_rating DESC
           LIMIT 5
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author  avg_rating  snt_rating
0                     Diana Gabaldon    4.300000          50
1         J.K. Rowling/Mary GrandPré    4.288462         312
2                    Agatha Christie    4.283019          53
3  Markus Zusak/Cao Xuân Việt Khương    4.264151          53
4                     J.R.R. Tolkien    4.240964         166

In [35]:
query = '''SELECT a.author, AVG (rat.rating) AS avg_rating, COUNT (DISTINCT rat.rating_id) AS snt_rating
           FROM authors AS a
           LEFT JOIN books AS b ON a.author_id = b.author_id
           LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
           GROUP BY a.author
           HAVING COUNT (DISTINCT rat.rating_id) >= 50
           
           ORDER BY snt_rating DESC
           LIMIT 5
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author  avg_rating  snt_rating
0  J.K. Rowling/Mary GrandPré    4.288462         312
1              J.R.R. Tolkien    4.240964         166
2             Stephenie Meyer    3.662500         160
3                   Dan Brown    3.741259         143
4             Nicholas Sparks    3.882883         111

Автор с самой высокой оценкой книг - это Diana Gabaldon (учитывая только книги с 50 и более оценками). Однако у J.K. Rowling/Mary GrandPré средняя оценка ниже всего на 0,01 пункта, но больше всего полученных оценок.

In [36]:
query = '''WITH book_ratings AS (SELECT rat.book_id,
                                        AVG(rat.rating) AS avg_rating,
                                        COUNT(rat.rating) AS cnt_ratings
                                 FROM ratings AS rat
                                 GROUP BY rat.book_id
                                 HAVING COUNT(rat.rating) >= 50),
                average_ratings AS (SELECT a.author_id,
                                             SUM(br.cnt_ratings) AS cnt_ratings,
                                             AVG(br.avg_rating) AS avg_rating
                                      FROM authors a
                                      JOIN books AS b ON a.author_id = b.author_id
                                      JOIN book_ratings AS br ON b.book_id = br.book_id
                                      GROUP BY a.author_id)
           SELECT a.author,
                  ar.cnt_ratings,
                  ar.avg_rating
           FROM authors a
           JOIN average_ratings AS ar ON a.author_id = ar.author_id
           ORDER BY ar.avg_rating DESC
           LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author  cnt_ratings  avg_rating
0         J.K. Rowling/Mary GrandPré        310.0    4.283844
1  Markus Zusak/Cao Xuân Việt Khương         53.0    4.264151
2                     J.R.R. Tolkien        162.0    4.258446
3                  Louisa May Alcott         52.0    4.192308
4                       Rick Riordan         62.0    4.080645

Автор с самой высокой оценкой книг - это J.K. Rowling (учитывая только книги с 50 и более оценками). И количество оценок у нее наибольшее.

**5.Среднее количество обзоров от пользователей, которые поставили больше 48 оценок**

In [37]:
query = '''SELECT ROUND(AVG(rev.count)) AS round
           FROM (SELECT COUNT(review_id)
                 FROM reviews 
                 WHERE username IN (SELECT  username
                              FROM ratings AS rat
                              GROUP BY username
                              HAVING COUNT (DISTINCT rat.rating_id) > 48
                              )
GROUP BY username
) AS rev
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

round
0   24.0

Пользователи, поставившие более 48 оценок, написали в среднем 24 обзора.

**Вывод**

1. После 1 января 2020 года выпущено 819 книг.

2. У 44 книг рейтинг равен 5, но количество обзоров у большинства из них 2. Больше всего отзывов (4) с рейтингом 5.0 у книги A Dirty Job (Grim Reaper #1). Больше всего отзывов у книги Twilight (Twilight #1)	 - 7.

3. Penguin Books - издательство, выпустившее наибольшее количество книг (42) толщиной более 50 страниц.

4. Автор с самой высокой оценкой книг - это J.K. Rowling (учитывая только книги с 50 и более оценками). И количество оценок у нее наибольшее.

5. Пользователи, поставившие более 48 оценок, написали в среднем 24 обзора.